### ISSUE TITLE
False positive: value typed as the union of a constrained TypeVar's constraints is not assignable back to the TypeVar

### ISSUE THREAD

**Issue by michael-bily** [astral-sh/ty#4162]
## Summary

When a `TypeVar` is constrained to `(A, B, A | B)` — i.e. the third constraint is the union of the first two — ty rejects two things that pyright and mypy accept:

1. Reassigning a variable of the TypeVar type with a value whose static type is `A | B` (e.g. the result of a `Self`-returning method), and
2. Passing an `A | B`-typed value into a function parameterised by that TypeVar.

Minimal reproducible example:

```python
from typing import TypeVar


class DF:
    def sort(self) -> "DF":
        return self


class LF:
    def sort(self) -> "LF":
        return self


# Third constraint is the union of the first two.
T = TypeVar("T", DF, LF, "DF | LF")


def reassign(x: T) -> T:
    x = x.sort()   # error[invalid-assignment]: `DF | LF` is not assignable to `T`
    return x


def pass_union(v: "DF | LF") -> None:
    reassign(v)    # error[invalid-argument-type]: `DF | LF` not accepted for `T`
```

Command: `ty check`

`ty` reports 3 diagnostics:

- `invalid-assignment: Object of type DF | LF is not assignable to T@reassign`
- `invalid-argument-type: Argument to bound method DF.sort is incorrect` (Expected `DF`, found `T`)
- `invalid-argument-type: Argument to bound method LF.sort is incorrect` (Expected `LF`, found `T`)

Expected: no errors. Both pyright (0 errors, 0 warnings) and mypy 2.3.0 --strict (Success: no issues found) accept this file.

**Real-world context:** this pattern is common with polars, where a helper is generic over `DataFrame`/`LazyFrame` and their union:

```python
PlDataFrame = TypeVar("PlDataFrame", "pl.DataFrame", "pl.LazyFrame", "pl.DataFrame | pl.LazyFrame")
```

In a real codebase this single pattern produces a few hundred false positives (every `df = df.sort(...)`-style reassignment and every call passing an un-narrowed `DataFrame | LazyFrame`). Dropping the union constraint isn't a workaround — it then (correctly) rejects callers that hold the union, and pyright regresses too.

Seems related to the constrained-TypeVar solving tracked under #623, but I couldn't find an existing issue for this specific direction (a union-of-constraints value being rejected). Happy to be pointed at a dupe if there is one.

Thanks for the great work on ty — it's already catching plenty and is a joy to run.

### Version

ty 0.0.65 (87de836df 2026-07-29)

**Comment by carljm** [comment]
Thanks for the report! This is #1503 (similar to #3605).
